### PR TITLE
feat: extend action space

### PR DESCRIPTION
**Pull request by strakam**
Extended action space, tested.
Expander maybe more readable.

### CODE REVIEW

File: generals/utils.py [hunk 1]
```diff
@@ -132,15 +130,15 @@ def store_replay(
...
-                ]
-            )
-            f.write(f"{row}\n")
+        for player_actions in action_sequence:
+            player_string = {}
+            for player in players:
+                act = player_actions[player]
+                player_string[player] = (
+                    f"{player}:{act[0]} {act[1]} {act[2]} {act[3]} {act[4]}"
+                )
```
Comment (jdujava) [inline generals/utils.py:139]: Can't this be done more elegantly, for arbirtary length of `act`?

File: generals/agents.py [hunk 2]
```diff
@@ -28,13 +30,15 @@ def play(self, observation):
...
         """
-        mask = observation['action_mask']
+        mask = observation["action_mask"]
         valid_actions = np.argwhere(mask == 1)
         action_index = np.random.choice(len(valid_actions))
-        # append 1 or 0 randomly to the action (to say whether to send half of troops or all troops)
-        action = np.append(valid_actions[action_index], np.random.choice([0, 1]))
+
+        # [1] to indicate we want to move, [0] to indicate we want to move all troops
+        action = np.concatenate(([1], valid_actions[action_index], [0]))
```
Comment (jdujava) [inline generals/agents.py:38]: I liked that `RandomClient` can sometimes move only half of troops, but maybe half of the time is really too much.

File: examples/gymnasium_example.py [hunk 3]
```diff
@@ -15,10 +15,10 @@
 
 # Create environment
 env = gym_generals(game_config, render_mode="human") # render_mode {"none", "human"}
-observation, info = env.reset(options={"replay_file": "test"})
+observation, info = env.reset()
```
Comment (jdujava) [inline examples/gymnasium_example.py:18]: Why disable replay?

File: examples/pettingzoo_example.py [hunk 4]
```diff
@@ -18,7 +18,7 @@
 
 # Create environment
 env = pz_generals(game_config, render_mode="human") # render_mode {"none", "human"}
-observations, info = env.reset(options={"replay_file": "test"})
+observations, info = env.reset()
```
Comment (jdujava) [inline examples/pettingzoo_example.py:21]: Why disable reaplay?

**Review by jdujava — COMMENTED** [review]
I like it, just have some questions.

File: generals/agents.py (hunk 2 above)
Comment (strakam) [reply generals/agents.py:38]: will fix as discussed.

File: generals/game.py [hunk 5]
```diff
@@ -56,27 +56,24 @@ def __init__(self, map: np.ndarray, agents: List[str]):
         ##########
         # Spaces #
         ##########
-        box = gym.spaces.Box(low=0, high=1, shape=spatial_dim, dtype=np.float32)
+        grid_multi_binary = gym.spaces.MultiBinary(spatial_dim)
         self.observation_space = gym.spaces.Dict(
             {
                 "army": gym.spaces.Box(
                     low=0, high=np.inf, shape=spatial_dim, dtype=np.float32
                 ),
```
Comment (jdujava) [inline generals/game.py:64]: Shouldn't `dtype` be `int` or something like that? Similarly in numerous other places like `test_game.py`, `rendering.py`, and in this file?

**Review by jdujava — COMMENTED** [review]
Just comment about typing.

File: generals/game.py (hunk 5 above)
Comment (strakam) [reply generals/game.py:64]: I'll think about it. The thing is that this value is then used for ML frameworks who convert it back to `float32` anyway. Probably it is a good convention to make it int.

**Pull request merged**